src/ledger/LedgerStateSnapshot.h

```diff
@@ -0,0 +1,161 @@
+// Copyright 2024 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#pragma once
+
+#include "bucket/BucketListSnapshot.h"
+#include "bucket/BucketSnapshotManager.h"
+#include "ledger/LedgerTxn.h"
+#include "util/NonCopyable.h"
+
+namespace stellar
+{
+
+class Application;
+class TransactionFrame;
+class LedgerSnapshot;
+
+// A unified ledger entry interface that supports LedgerEntry representations
+// for both legacy SQL and BucketList snapshots. When working with LedgerTxn,
+// using plain LedgerEntry is not safe, as it might be modified or invalidated
+// by a nested transaction. To address this, the LedgerTxnEntry and
+// ConstLedgerTxnEntry abstractions should be preserved, allowing bucket
+// snapshot LedgerEntries to support the same interface.
+class LedgerEntryWrapper
+{
+    // Either hold a reference or a pointer to the entry
+    std::variant<LedgerTxnEntry, ConstLedgerTxnEntry,
+                 std::shared_ptr<LedgerEntry>>
+        mEntry;
+
+  public:
+    LedgerEntryWrapper(ConstLedgerTxnEntry&& entry);
+    LedgerEntryWrapper(LedgerTxnEntry&& entry);
+    LedgerEntryWrapper(std::shared_ptr<LedgerEntry> entry);
+    LedgerEntry const& current() const;
+    operator bool() const;
+};
+
+// A unified ledger header access interface, similar to
+// LedgerEntryWrapper. Just like LedgerEntryWrapper, this class is purely
+// cosmetic for BucketList snapshots, since those are immutable.
+class LedgerHeaderWrapper
+{
+    std::variant<LedgerTxnHeader, std::shared_ptr<LedgerHeader>> mHeader;
+
+  public:
+    LedgerHeaderWrapper(LedgerTxnHeader&& header);
+    LedgerHeaderWrapper(LedgerHeader const& header);
+    LedgerHeaderWrapper(std::shared_ptr<LedgerHeader> header);
+    LedgerHeader& currentToModify();
+    LedgerHeader const& current() const;
+    LedgerTxnHeader const&
+    getLedgerTxnHeader() const
+    {
+        releaseAssert(std::holds_alternative<LedgerTxnHeader>(mHeader));
+        return std::get<0>(mHeader);
+    }
+};
+
+// A unified interface for read-only ledger state snapshot.
+// Supports SQL (via read-only LedgerTxn), as well as BucketList snapshots.
+class AbstractLedgerStateSnapshot
+{
+  public:
+    virtual ~AbstractLedgerStateSnapshot() = default;
+    virtual LedgerHeaderWrapper getLedgerHeader() const = 0;
+    // Get ledger header associated with this snapshot. This is not always safe
+    // to do, as this method does not do any validation, if a nested snapshot
+    // exists (which invalidates the header).
```

Comment:
Is it unsafe only within `executeWithMaybeInnerSnapshot` or are there some other scenarios when this might happen?

Reply:
This is never safe to do - we should be properly using `loadHeader` that checks invariants. A bit more info in this [issue](https://github.com/stellar/stellar-core/issues/4412)